vowpalwabbit/cb_explore_adf_synthcover.cc

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "cb_explore_adf_synthcover.h"
+
+#include "cb_explore_adf_common.h"
+#include "reductions.h"
+#include "cb_adf.h"
+#include "rand48.h"
+#include "bs.h"
+#include "gen_cs_example.h"
+#include "cb_explore.h"
+#include "explore.h"
+#include "vw_versions.h"
+#include "version.h"
+#include <vector>
+#include <algorithm>
+#include <cmath>
+
+// All exploration algorithms return a vector of id, probability tuples, sorted in order of scores. The probabilities
+// are the probability with which each action should be replaced to the top of the list.
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+namespace synthcover
+{
+struct cb_explore_adf_synthcover
+{
+private:
+  float _epsilon;
+  float _psi;
+  size_t _synthcoversize;
+  std::shared_ptr<rand_state> _random_state;
+
+  VW::version_struct _model_file_version;
+
+  v_array<ACTION_SCORE::action_score> _action_probs;
+  std::vector<int> policyaction;
+  float min_cost;
+  float max_cost;
+
+public:
+  cb_explore_adf_synthcover(float epsilon, float psi, size_t synthcoversize, std::shared_ptr<rand_state> random_state,
+      VW::version_struct model_file_version);
+  ~cb_explore_adf_synthcover();
+
+  // Should be called through cb_explore_adf_base for pre/post-processing
+  void predict(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<false>(base, examples); }
+  void learn(VW::LEARNER::multi_learner& base, multi_ex& examples) { predict_or_learn_impl<true>(base, examples); }
+  void save_load(io_buf& model_file, bool read, bool text);
+
+private:
+  template <bool is_learn>
+  void predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples);
+};
+
+cb_explore_adf_synthcover::cb_explore_adf_synthcover(float epsilon, float psi, size_t synthcoversize,
+    std::shared_ptr<rand_state> random_state, VW::version_struct model_file_version)
+    : _epsilon(epsilon)
+    , _psi(psi)
+    , _synthcoversize(synthcoversize)
+    , _random_state(random_state)
+    , _model_file_version(model_file_version)
+    , _action_probs(v_init<ACTION_SCORE::action_score>())
+    , min_cost(0.0)
```

Comment:
are costs ever negative? If not, doesn't `min_cost` need to be initialized to a max numeric limit instead of `0.0` in the constructor? Otherwise it seems it will never get replaced.

Reply:
It's ok to assume that a cost of 0 was observed at the beginning of time. I just need to clip the underlying prediction to something reasonable, otherwise the exploration bonus can get overwhelmed by large predictions.